### PR TITLE
Issue 606  quadratic form functional

### PR DESCRIPTION
**Pull request by adler-j**

### CODE REVIEW

File: odl/solvers/functional/default_functionals.py [hunk 1]
```diff
@@ -570,6 +570,139 @@ def __init__(self, space):
...
 
+class QuadraticForm(Functional):
+
+    """The functional for a general quadratic form ``x^T A x + b^T x + c``."""
+
+    def __init__(self, operator=None, offset=None, constant=0):
+        """Initialize a new instance.
+
+        All parameters are optional, but at least one of ``op`` and ``offset``
+        have to be provided in turn to infer the space.
```
Comment (kohr-h) [inline odl/solvers/functional/default_functionals.py]: in order to

File: odl/solvers/functional/default_functionals.py [hunk 2]
```diff
@@ -570,6 +570,139 @@ def __init__(self, space):
         IndicatorBox.__init__(self, space, lower=0, upper=None)
 
 
+class QuadraticForm(Functional):
+
+    """The functional for a general quadratic form ``x^T A x + b^T x + c``."""
+
+    def __init__(self, operator=None, offset=None, constant=0):
```
Comment (kohr-h) [inline odl/solvers/functional/default_functionals.py]: I would think that `offset` is `c`. Go for a neutral term like `vector`?

File: odl/solvers/functional/default_functionals.py [hunk 3]
```diff
@@ -570,6 +570,139 @@ def __init__(self, space):
...
+        have to be provided in turn to infer the space.
+
+        The computed value is::
+
+            x.inner(operator(x)) + offset.inner(x) + constant
+
+        Parameters
+        ----------
+        operator : `Operator`, optional
+            Linear Operator for the quadratic part of the functional.
```
Comment (kohr-h) [inline odl/solvers/functional/default_functionals.py]: operator

File: odl/solvers/functional/default_functionals.py [hunk 4]
```diff
@@ -570,6 +570,139 @@ def __init__(self, space):
...
+            ``None`` means this is ignored.
```
Comment (kohr-h) [inline odl/solvers/functional/default_functionals.py]: this -> that this part

File: odl/solvers/functional/default_functionals.py [hunk 5]
```diff
@@ -570,6 +570,139 @@ def __init__(self, space):
...
+        offset : `Operator`, optional
+            Vector for the linear part of the functional.
+            ``None`` means this is ignored.
```
Comment (kohr-h) [inline odl/solvers/functional/default_functionals.py]: same here

File: odl/solvers/functional/default_functionals.py [hunk 6]
```diff
@@ -570,6 +570,139 @@ def __init__(self, space):
...
+        constant : `Operator`, optional
+            Constant offset of the functional.
+        """
+        if operator is None and offset is None:
+            raise ValueError('need to provide at least one of ``op`` and '
```
Comment (kohr-h) [inline odl/solvers/functional/default_functionals.py]: operator

File: odl/solvers/functional/default_functionals.py [hunk 7]
```diff
@@ -570,6 +570,139 @@ def __init__(self, space):
...
+            return ConstantOperator(self.domain, self.offset)
+        else:
+            if not self.operator.is_linear:
+                # TODO: Acutally works otherwise, but needs more work
+                raise NotImplementedError('`operator` must be linear')
+
+            # Figure out if operator is symmetric
+            opadjoint = self.operator.adjoint
+            if opadjoint == self.operator:
+                opgradient = 2 * self.operator
```
Comment (kohr-h) [inline odl/solvers/functional/default_functionals.py]: Why **op**gradient?

File: odl/solvers/functional/default_functionals.py [hunk 8]
```diff
@@ -570,6 +570,139 @@ def __init__(self, space):
...
+            return x.inner(tmp) + self.constant
+
+    @property
+    def gradient(self):
+        """Gradient operator of the functional."""
+        if self.operator is None:
+            return ConstantOperator(self.domain, self.offset)
+        else:
+            if not self.operator.is_linear:
+                # TODO: Acutally works otherwise, but needs more work
```
Comment (kohr-h) [inline odl/solvers/functional/default_functionals.py]: It is for `F(x) = <A(x), x> + <b, x> + c`
```
d/dt_t=0 F(x + ty) = <A'(x) y, x> + <A(x), y> + <b, y> = <A'(x)^* x + A(x) + b, y>
```
i.e. the gradient is `A'(x)^* x + A(x) + b`.

Comment (adler-j) [inline odl/solvers/functional/default_functionals.py]: Yeah, i know but try writing that in some sensible manner. The operator `x -> A'(x)^*(x)` cannot be written in a nice way so I'm leaving it for now.

**Comment by adler-j** [comment]
Merge after CI

**Pull request merged**